packages/monitor-v2/src/balance-monitor/BalanceMonitor.ts

```diff
@@ -0,0 +1,153 @@
+import { BigNumber, utils } from "ethers";
+
+import { getContractInstanceWithProvider, getCurrencyDecimals, getCurrencySymbol } from "../utils/contracts";
+
+export interface MonitoredToken {
+  address?: string; // If not provided, this is monitoring native token.
+  warnThreshold?: number; // Human readable balance threshold to trigger a warning.
+  errorThreshold?: number; // Human readable balance threshold to trigger an error.
+}
+
+export interface MonitoredAccount {
+  chainId: number;
+  address: string;
+  name?: string; // If not provided, address is used as name.
+  tokens: MonitoredToken[];
+}
+
+// Internal interface to hold monitored item in the BalanceMonitor.
+interface MonitoredBalance {
+  accountName: string;
+  accountAddress: string;
+  provider: Provider;
+  chainId: number;
+  networkName: string;
+  tokenSymbol: string;
+  tokenAddress?: string; // If not provided, this is monitoring native token.
+  tokenDecimals: number;
+  warnThreshold: BigNumber; // Raw balance threshold to trigger a warning.
+  errorThreshold: BigNumber; // Raw balance threshold to trigger an error.
+}
+
+function isMonitoredToken(obj: unknown): obj is MonitoredToken {
+  if (typeof obj !== "object" || obj === null) return false;
+  const token = obj as MonitoredToken;
+
+  const hasValidAddress =
+    token.address === undefined || (typeof token.address === "string" && utils.isAddress(token.address));
+  const hasValidWarnThreshold = token.warnThreshold === undefined || typeof token.warnThreshold === "number";
+  const hasValidErrorThreshold = token.errorThreshold === undefined || typeof token.errorThreshold === "number";
+
+  return hasValidAddress && hasValidWarnThreshold && hasValidErrorThreshold;
+}
+
+function isMonitoredAccount(obj: unknown): obj is MonitoredAccount {
+  if (typeof obj !== "object" || obj === null) return false;
+  const account = obj as MonitoredAccount;
+
+  const hasValidChainId = typeof account.chainId === "number";
+  const hasValidAddress = typeof account.address === "string" && utils.isAddress(account.address);
+  const hasValidName = account.name === undefined || typeof account.name === "string";
+  const hasValidTokens = Array.isArray(account.tokens) && account.tokens.every(isMonitoredToken);
+
+  return hasValidChainId && hasValidAddress && hasValidName && hasValidTokens;
+}
+
+export function isMonitoredAccountsArray(obj: unknown): obj is MonitoredAccount[] {
+  if (!Array.isArray(obj)) return false;
+  return obj.every(isMonitoredAccount);
+}
+
+export class BalanceMonitor {
+  private constructor(private readonly monitoredBalances: MonitoredBalance[]) {}
+
+  // Internal method to get raw threshold from human readable amount.
+  private static parseThreshold(threshold: number | undefined, decimals: number): BigNumber {
+    if (threshold === undefined) return BigNumber.from(0);
+    return utils.parseUnits(threshold.toString(), decimals);
+  }
+
+  static async create(
```

Comment:
+1 on the static create function. Allows us to have a concrete, unchanging list in the class!